core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -243,6 +243,22 @@ class KafkaApis(val requestChannel: RequestChannel,
         s"${updateMetadataRequest.brokerEpoch()} smaller than the current broker epoch ${controller.brokerEpoch}")
       sendResponseExemptThrottle(request, new UpdateMetadataResponse(Errors.STALE_BROKER_EPOCH))
     } else {
+      def shouldUpdateInternalTopicPartitionCount(topicName: String) = {
+        val partitionsOfTopic = updateMetadataRequest.partitionStates.keySet.asScala.filter(_.topic == topicName)
+        if (partitionsOfTopic.isEmpty)
+          false
+        else {
+          val maxPartitionCountInRequest = partitionsOfTopic.map(_.partition).max
+          !metadataCache.getPartitionInfo(topicName, maxPartitionCountInRequest).isDefined
+        }
+      }
+
+      if (shouldUpdateInternalTopicPartitionCount(GROUP_METADATA_TOPIC_NAME))
```

Comment:
We should be comparing whether the count known to the coordinator (e.g. what is returned by `getGroupMetadataTopicPartitionCount`) matches this count we have from the updated metadata. Maybe we can have a method like `GroupCoordinator.maybeRefreshMetadataTopic(partitionId: Int)` or something.